python/tvm/relax/frontend/nn/core.py

```diff
@@ -591,7 +609,22 @@ def wrap_nested(expr: rx.Expr, name: str) -> Union[Tensor, Sequence[Tensor]]:
         The computed result.
     """
     if not isinstance(expr, rx.DataflowVar):
-        expr = BlockBuilder.current().emit(expr, name)
+        block_builder = BlockBuilder.current()
+        if block_builder is None:
+            # Normalize to make sure we have valid StructInfo, but
+            # wait until we are actually building the function to
+            # flatten nested expressions.
+            #
+            # TODO(Lunderberg): Make this easier to call.  Infering
+            # struct info for a nested expression should be doable in
+            # a free function, without requiring an active
+            # BlockBuilder and an active FunctionFrame.
```

Comment:
Long-term, I think it would be nice to distinguish between local struct inference and non-local struct inference.  The local inference could be applied when a relax object is constructed, which would avoid the current two-phase initialization of relax objects.  Since this step can only perform local struct inference, which would be applied by default, this entire conditional could be removed.

There's some kinks that would need to be worked out first.  Some of the struct inference for tensor operations currently throw errors a bit more than I think they should.   (e.g. If `R.matmul` throws an exception if the arguments are not `R.Tensor`.  If the arguments are `R.Object`, the exception is still thrown, even though `R.Tensor` is a subtype of `R.Object`.)  These fallbacks would probably get more exercise with local inference, as there may be less information available.